File: zemiao/211.py
class TrieNode(object):
  def __init__(self):
    self.vals = [None]*26
    self.isLeaf = False
    
  def search(self,word):
    if not word:
      return self.isLeaf
    c = word[0]
    if c == '.':
      for node in self.vals:
        if not node:
          continue
        if node.search(word[1:]):
          return True
      return False
    if not self.vals[ord(c)-ord('a')]:
      return False
    else:
      return self.vals[ord(c)-ord('a')].search(word[1:])

class WordDictionary(object):

  def __init__(self):
    """
    Initialize your data structure here.
    """
    self.root = TrieNode()

  def addWord(self, word):
    """
    Adds a word into the data structure.
    :type word: str
    :rtype: void
    """
    node = self.root
    for c in word:
      if not node.vals[ord(c)-ord('a')]:
        node.vals[ord(c)-ord('a')] = TrieNode()
      node = node.vals[ord(c)-ord('a')] 
    node.isLeaf = True

  def search(self, word):
    """
    Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
    :type word: str
    :rtype: bool
    """
    return self.root.search(word)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: tnzhou/211.cpp
class WordDictionary {
public:
    WordDictionary *child[26];
    bool wordEnd;

    /** Initialize your data structure here. */
    WordDictionary() {
        for (int i = 0; i < 26; i ++)
            child[i] = nullptr;
        wordEnd = false;
    }
    
    /** Adds a word into the data structure. */
    void addWord(string word) {
        WordDictionary *curNode = this;
        for (int i = 0; i < word.length(); i++) {
            int cIndex = word[i] - 'a';
            if (curNode->child[cIndex] == nullptr) {
                curNode->child[cIndex] = new WordDictionary();
            }
            curNode = curNode->child[cIndex];
        }
        curNode->wordEnd = true;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        return dfsSearch(this, word);
    }

private:
    bool dfsSearch(WordDictionary *curNode, string restStr) {
        if (restStr.length() == 0)
            return curNode->wordEnd;
        if (restStr[0] == '.') {
            // All way search
            string remainedStr = restStr.substr(1);
            for (int i = 0; i < 26; i++) {
                if (curNode->child[i] != nullptr && dfsSearch(curNode->child[i], remainedStr))
                    return true;
            }
        } else {
            int cIndex = restStr[0] - 'a';
            if (curNode->child[cIndex] == nullptr) {
                return false;
            } else {
                return dfsSearch(curNode->child[cIndex], restStr.substr(1));
            }
        }
        return false;
    }    
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */
